### PR TITLE
Optimize semaphore acquisition in GpuShuffledHashJoinExec

### PR DESCRIPTION
**Pull request by abellina**
Closes https://github.com/NVIDIA/spark-rapids/issues/4539.

This PR adds an optimization for the shuffled hash join where the semaphore is not taken after the build side is fetched to the host (since we attempt to keep the build side on the host while we load the first batch from the stream side).

After the first stream batch is loaded, the semaphore has been acquired. We then proceed to bring the build batch to the GPU.

This code falls back to the old behavior in cases where batches are not serialized (our input is not a shuffle) and when the streams are empty. 

It is in draft mode because I need to run more testing in NDS. I don't think the case where I need to loop over the build batch multiple times has been hit with my runs so far (other than unit tests), so I need to figure out a way to trigger that case.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 1]
```diff
@@ -129,22 +133,22 @@ case class GpuShuffledHashJoinExec(
...
       (streamIter, buildIter) => {
-        val stIt = new CollectTimeIterator("shuffled join stream", streamIter, streamTime)
-        val startTime = System.nanoTime()
-
-        withResource(ConcatAndConsumeAll.getSingleBatchWithVerification(buildIter,
-          localBuildOutput)) { builtBatch =>
+        val (builtBatch, maybeBufferedStreamIter) =
+          GpuShuffledHashJoinExec.getBuiltBatchAndStreamIter(
+            targetSize,
+            buildPlan.output,
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:142]: Go back to `localBuildOutput`. It is there to avoid needing to serialize the entire buildPlan when shipping the zipPartitions function.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala [hunk 2]
```diff
@@ -237,6 +237,27 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
       p.withNewChildren(p.children.map(optimizeCoalesce))
   }
 
+  /**
+   * Removes `GpuCoalesceBatches(GpuShuffleCoalesceExec(build side))` for the build side
+   * for the shuffled hash join. The coalesce logic has been moved to the
+   * `GpuShuffleCoalesceExec` class, and is handled differently to prevent holding onto the
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala:243]: This is fine for now, but I really would prefer to have us just build the plan the right way from the beginning.  Perhaps we can have a flag in GpuExec that says for this input I know how to handle raw shuffle data so then the rules that insert the host side coalesce can deal with it there.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 3]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+   * @param buildTime build GPU metric
+   * @return a pair of `ColumnarBatch` and streamed iterator that can be
+   *         used for the join
+   */
+  def getBuiltBatchAndStreamIter(targetBatchSize: Long,
+                                 buildOutput: Seq[Attribute],
+                                 buildIter: Iterator[ColumnarBatch],
+                                 streamIter: Iterator[ColumnarBatch],
+                                 semWait: GpuMetric,
+                                 buildTime: GpuMetric): (ColumnarBatch, Iterator[ColumnarBatch]) = {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:217]: nit: if it is simple it would be great to add some more metrics here for the operations that we combined into this.  We no longer know how long it took to concat on the GPU. At a minimum that should be added to the opTime metric.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 4]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+    closeOnExcept(bufferedBuildIterator) { _ =>
+      val startTime = System.nanoTime()
+      // find if the build side is non-empty, and if the first batch is
+      // a serialized batch. If neither condition is met, we fallback to the
+      // `getSingleBatchWithVerification` method.
+      val firstBatchIsSerialized = {
+        if (!buildIter.hasNext) {
+          false
+        } else {
+          bufferedBuildIterator = new CloseableBufferedIterator(buildIter.buffered)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:228]: How much is buffered when we do this?

**Comment by revans2** [comment]
So just to be clear this optimization can never work with UCX right?

**Comment by abellina** [comment]
> So just to be clear this optimization can never work with UCX right?

Great point. I have missed a case where we need to ignore the config when UCX is turned on and we have compression on. If we don't do this, the fallback case right now doesn't handle compressed batches from the shuffle.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 5]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+   * @param targetBatchSize configured target batch size
+   * @param buildOutput output attributes of the build plan
+   * @param buildIter build iterator
+   * @param streamIter stream iterator
+   * @param semWait wait GPU metric
+   * @param buildTime build GPU metric
+   * @return a pair of `ColumnarBatch` and streamed iterator that can be
+   *         used for the join
+   */
+  def getBuiltBatchAndStreamIter(targetBatchSize: Long,
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:212]: We cannot as any of the iterators if they `haveNext` unless we hold the GPU semaphore or we know 100% that they will not put something on the GPU.  I the general case we cannot know that. So we cannot call it on the stream iterator and we cannot call it on the build iterator to discover what type of iterator it is.

**Review by revans2 — CHANGES_REQUESTED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 5 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:212]: I might not understand how/when the semaphore is grabbed in these cases. If I am wrong, which I hope I am, then we need some more comments explaining how this all works.

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:212]: Ya I rethought it. Can we have some comments around the hasNext calls saying that they will grab the semaphore if they are going to put something on the GPU and we don't need to worry about it in those cases.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 6]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+      }
+
+      if (!firstBatchIsSerialized || !streamIter.hasNext) {
+        // fallback if we failed to find serialized build batches, or if the
+        // stream iterator is empty
+        val builtBatch =
+        ConcatAndConsumeAll.getSingleBatchWithVerification(
+          Option(bufferedBuildIterator).getOrElse(buildIter), buildOutput)
+        val delta = System.nanoTime() - startTime
+        buildTime += delta
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:248]: So does the build time in this case also include the time to check to see if the streamIter is empty? If so I am fine with it, we just might want to document it because it could be really confusing.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 5 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:212]: Yeah, if hasNext is called in the buffered iterator a single batch is getting loaded. That batch may be on the GPU. The expectation is that the iterator that gave us that batch via the buffered iterator is getting on the GPU, if not that's a bug. I'll put a comment around this to make this clear.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 7]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+        (builtBatch, streamIter)
+      } else {
+        val dataTypes = buildOutput.map(_.dataType).toArray
+        val bufferedStreamIter = new CloseableBufferedIterator(streamIter.buffered)
+        val hostBuffs = new ArrayBuffer[SerializedTableColumn]()
+        var buildBatch: ColumnarBatch = null
+        closeOnExcept(bufferedStreamIter) { _ =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:255]: nit: why make this two lines of code?

```
closeOnExcept(new CloseableBufferedIterator(streamIter.buffered)) { bufferedStreamIter =>
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 8]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+                while (bufferedBuildIterator.hasNext && sizeInHost <= targetBatchSize) {
+                  val hostBatch = bufferedBuildIterator.next()
+                  val serializedTableColumn =
+                    hostBatch.column(0).asInstanceOf[SerializedTableColumn]
+                  hostBuffs += serializedTableColumn
+                  sizeInHost += serializedTableColumn.hostBuffer.getLength
+                }
+
+                // concat our host batches on the host
+                val hostConcat = withResource(hostBuffs) { _ =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:269]: We need to move the close block for hostConcat higher up. There is a require in between hostConcat being created and when it is safe to throw an exception.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 9]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+        // stream iterator is empty
+        val builtBatch =
+        ConcatAndConsumeAll.getSingleBatchWithVerification(
+          Option(bufferedBuildIterator).getOrElse(buildIter), buildOutput)
+        val delta = System.nanoTime() - startTime
+        buildTime += delta
+        (builtBatch, streamIter)
+      } else {
+        val dataTypes = buildOutput.map(_.dataType).toArray
+        val bufferedStreamIter = new CloseableBufferedIterator(streamIter.buffered)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:252]: Do we really need a buffered iterator on the stream-side batch?  We just need to call `hasNext` which will grab the semaphore but I don't see the need for us to call `next` on the stream-side iterator.  That also side-steps another bug here where we create a closeable buffered iterator but pass it to code that will never call close on it since it's just a regular iterator to that code.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 10]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+        val hostBuffs = new ArrayBuffer[SerializedTableColumn]()
+        var buildBatch: ColumnarBatch = null
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:254]: I'm not seeing the need to declare this at this relatively high scope.  IIUC there are two basic cases here:
- We hit the end of the iterator before the target size, we can then build the batch immediately and return it, so no need to track this batch outside of the while loop.
- We hit the target size before the end of the iterator, which then we need to coalesce these all together into a single batch.  That can be accomplished by creating a new iterator by tacking our host-concat-and-put-on-GPU batch on the front of a `GpuShuffleCoalesceIterator` that wraps the build iterator we were using.  We then pass this to `ConcatAndConsumeAll.getSingleBatchWithVerification` and return it immediately.

In other words, this while loop should never be building batches on the GPU iteratively, only fetching host batches until one of two things happen which will be the last iteration of this while loop.

**Comment by abellina** [comment]
I added a commit that I think addresses many of the points, enough has changed that I think this needs another pass.

One comment from Jason wasn't addressed for sure:

> Do we really need a buffered iterator on the stream-side batch? We just need to call hasNext which will grab the semaphore but I don't see the need for us to call next on the stream-side iterator. That also side-steps another bug here where we create a closeable buffered iterator but pass it to code that will never call close on it since it's just a regular iterator to that code.

We need _something_ to hold on to that first stream batch. That `doJoin` pays no attention to this, so if there is an exception before that the first batch is popped, then the batch will be leaked.

I think one ugly option is to attach the closeable iterator to the task completion logic. I think we discussed doing this, but I am confirming that's the approach or if there's something better.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala [hunk 11]
```diff
@@ -46,15 +46,28 @@ object ConcatAndConsumeAll {
    * @return a single batch with all of them concated together.
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:46]: I should be good now.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala]: done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 6 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:258]: Ok I still have this. Let me handle.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:258]: This is no longer an issue.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 7 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala]: Done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 12]
```diff
@@ -470,6 +470,15 @@ object RapidsConf {
     .booleanConf
     .createWithDefault(false)
 
+  val OPTIMIZE_SHUFFLED_HASH_JOIN = conf("spark.rapids.sql.optimizeShuffledHashJoin")
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala]: Changed the config name an the function name for the optimization.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 13]
```diff
@@ -129,22 +133,22 @@ case class GpuShuffledHashJoinExec(
     val joinOutputRows = gpuLongMetric(JOIN_OUTPUT_ROWS)
     val targetSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(conf)
     val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
-    val localBuildOutput: Seq[Attribute] = buildPlan.output
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:132]: Yeap back to `localBuildOutput`.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 14]
```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
...
+          } else {
+            firstBatch.column(0) match {
+              case _: SerializedTableColumn => true
+              case _ => false
+            }
+          }
+        }
+      }
+
+      if (!firstBatchIsSerialized || !streamIter.hasNext) {
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala]: Yes absolutely. This should be fixed.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 15]
```diff
@@ -123,28 +127,35 @@ case class GpuShuffledHashJoinExec(
...
-    val buildTime = gpuLongMetric(BUILD_TIME)
     val streamTime = gpuLongMetric(STREAM_TIME)
     val joinTime = gpuLongMetric(JOIN_TIME)
     val joinOutputRows = gpuLongMetric(JOIN_OUTPUT_ROWS)
-    val targetSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(conf)
+    val batchSizeBytes = RapidsConf.GPU_BATCH_SIZE_BYTES.get(conf)
     val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
-    val localBuildOutput: Seq[Attribute] = buildPlan.output
+    val localBuildOutput = buildPlan.output
+    // stomp on metrics that don't make sense in the join
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:136]: This comment is very confusing.  All of these metrics are very relevant for joins.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 16]
```diff
@@ -155,3 +166,160 @@ case class GpuShuffledHashJoinExec(
...
+object GpuShuffledHashJoinExec extends Arm {
+  /**
+   * Helper iterator that wraps a BufferedIterator of AutoCloseable subclasses.
+   * This iterator also implements AutoCloseable, so it can be closed in case
+   * of exceptions.
+   *
+   * @param wrapped the buffered iterator
+   * @tparam T an AutoCloseable subclass
+   */
+  class CloseableBufferedIterator[T <: AutoCloseable](wrapped: BufferedIterator[T])
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:179]: Does this need a task completion listener?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 17]
```diff
@@ -155,3 +166,160 @@ case class GpuShuffledHashJoinExec(
...
+        } else {
+          bufferedBuildIterator = new CloseableBufferedIterator(buildIter.buffered)
+          val firstBatch = bufferedBuildIterator.head
+          if (firstBatch.numCols() != 1) {
+            false
+          } else {
+            firstBatch.column(0) match {
+              case _: SerializedTableColumn => true
+              case _ => false
+            }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:247]:
```suggestion
            firstBatch.column(0).isInstanceOf[SerializedTableColumn]
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 18]
```diff
@@ -155,3 +166,160 @@ case class GpuShuffledHashJoinExec(
...
+        buildTime += delta
+        (builtBatch, streamIter)
+      } else {
+        val dataTypes = buildOutput.map(_.dataType).toArray
+        closeOnExcept(new CloseableBufferedIterator(streamIter.buffered)) { bufferedStreamIter =>
+          val hostConcatIter = new MaybeHostConcatResultIterator(bufferedBuildIterator,
+              hostTargetBatchSize, dataTypes, coalesceMetricsMap)
+          closeOnExcept(hostConcatIter.next()) { maybeHostConcatResult =>
+            val buildBatch = if (!hostConcatIter.hasNext()) {
+              val singleBatchBufferTimeDelta = System.nanoTime() - startTime
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:267]: It's difficult to follow code when there's a `val somevar = if (...) {` followed by screens of code for the `if` body.  Suggest factoring this into a private method that is responsible for getting the build batch, whether that's using the single host buffer or falling back.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 19]
```diff
@@ -155,3 +166,160 @@ case class GpuShuffledHashJoinExec(
...
+            }
+          }
+        }
+      }
+
+      if (!firstBatchIsSerialized) {
+        // fallback if we failed to find serialized build batches
+        val builtBatch =
+          ConcatAndConsumeAll.getSingleBatchWithVerification(
+            Option(bufferedBuildIterator).getOrElse(buildIter), buildOutput)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:256]: There should be no need to emulate a fallback here.  If the first batch isn't serialized then that means we're not a join that was directly connected to a shuffle on the build-side.  That means the coalesce should have already occurred, otherwise there's something wrong with the planning.  May be prudent to `assert(!buildIter.hasNext)` but otherwise we should be able to just use the build batch and stream iter directly at this point.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala [hunk 20]
```diff
@@ -75,55 +113,76 @@ case class GpuShuffleCoalesceExec(child: SparkPlan, targetBatchByteSize: Long)
...
       while (batchCanGrow && iter.hasNext) {
         closeOnExcept(iter.next()) { batch =>
           inputBatchesMetric += 1
           // don't bother tracking empty tables
           if (batch.numRows > 0) {
-            inputRowsMetric += batch.numRows
             val tableColumn = batch.column(0).asInstanceOf[SerializedTableColumn]
             batchCanGrow = canAddToBatch(tableColumn.header)
             serializedTables.addLast(tableColumn)
+            inputRowsMetric += tableColumn.header.getNumRows
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala:185]: I thought I had reverted this change, but I didn't. Nice catch

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 15 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:136]: Making the comment better, thanks for pointing it out

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 16 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:179]: It's getting one!

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 19 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:256]: OK, sounds like a plan

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala [hunk 21]
```diff
@@ -75,55 +113,76 @@ case class GpuShuffleCoalesceExec(child: SparkPlan, targetBatchByteSize: Long)
  * to the target batch size and then concatenated on the host before the data
  * is transferred to the GPU.
  */
-class GpuShuffleCoalesceIterator(
+class MaybeHostConcatResultIterator(
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala]: This class went away.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala [hunk 22]
```diff
@@ -61,10 +61,48 @@ case class GpuShuffleCoalesceExec(child: SparkPlan, targetBatchByteSize: Long)
...
+        dataTypes, metricsMap)
+    }
+  }
+}
+
+/**
+ * A helper (but ugly) way of encapsulating the output of the
+ * `MaybeHostConcatResultIterator`. Batches fetched from the shuffle may or may
+ * not be empty, yet still on the host. There isn't a `HostConcatResult` for an empty
+ * batch, so this case class exists to expose a `getColumnarBatch` that always returns
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala]: This class went away.

File: sql-plugin/src/main/scala/ai/rapids/cudf/HostConcatResultUtil.scala [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+
+object HostConcatResultUtil extends Arm {
+  /**
+   * Create a `HostConcatResult` that only contains number of rows.
+   *
+   * In order to create an instance of `HostConcatResult` that is closeable,
+   * we allocate a 0-byte `HostMemoryBuffer`. Because this is requesting memory
+   * from the system, it becomes: UNSAFE.allocateMemory(0) which returns 0.
+   * The HostMemoryBuffer.close() logic skips such cases where the address
+   * is 0, so the resulting `HostConcatResult` can be used in withResources easily.
```
Comment (jlowe) [inline sql-plugin/src/main/scala/ai/rapids/cudf/HostConcatResultUtil.scala:33]: Nit: This is a lot of detail that IMO is unnecessary.  We're allocating an empty buffer, and we release that buffer when we're done.  How empty buffers are handled underneath is not this code's concern since it is adhering to the ownership convention of `HostMemoryBuffer`.

File: sql-plugin/src/main/scala/ai/rapids/cudf/HostConcatResultUtil.scala [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+      hostConcatResult: HostConcatResult,
+      sparkSchema: Array[DataType]): ColumnarBatch = {
+    if (hostConcatResult.getTableHeader.getNumColumns == 0) {
+      // We expect the caller to have acquired the GPU unconditionally before calling
+      // `getColumnarBatch`, as a downstream exec may need the GPU, and the assumption is
+      // that it is acquired in the coalesce code.
+      new ColumnarBatch(Array.empty, hostConcatResult.getTableHeader.getNumRows)
+    } else {
+      withResource(hostConcatResult.toContiguousTable) { ct =>
+        GpuColumnVector.from(ct.getTable, sparkSchema)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/ai/rapids/cudf/HostConcatResultUtil.scala:59]:
```suggestion
        GpuColumnVectorFromBuffer.from(ct, sparkSchema)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 19 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala]: Thinking about this one more, I am not seeing the issue, I am probably missing something. The call to `getSingleBatchWithVerification` is building an empty batch (if the build itereator is empty) and is checking that it gets a single GPU batch in the non-empty build iterator case.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 18 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala]: Ok this should be better here: https://github.com/NVIDIA/spark-rapids/pull/4588/commits/a607e73887ffd05ad863d1af4a464b2a3999ea4a.

**Review by jlowe — APPROVED** [review]

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
Looking into the test failures.

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
@jlowe @revans2 could you take another look when you have a chance?

**Review by jlowe — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala [hunk 25]
```diff
@@ -448,9 +458,84 @@ class GpuCoalesceIterator(iter: Iterator[ColumnarBatch],
     batches.append(SpillableColumnarBatch(batch, SpillPriorities.ACTIVE_BATCHING_PRIORITY,
       spillCallback))
 
+  protected def popAll(): Array[ColumnarBatch] = {
+    closeOnExcept(batches.map(_.getColumnarBatch())) { wip =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:462]: nit: does this need to be a safeMap?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 4 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:251]: Still would like to know the answer to this. The things in the iterator need to be closed and can potentially be very large.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:251]: When we call `.buffered` on the iterator it won't actively buffer anything, it's wrapping the original iterator in an interator that would buffer 1 `ColumnarBatch` in this case.

When we call `head` that batch is stored in the `BufferedIterator`. The `CloseableBufferedIterator` was added to actively close that if there is an exception.

**Comment by abellina** [comment]
@revans2 I replied to your comment just now in the original request, but adding a link here so we can continue if there's more I need to look into: 

> Still would like to know the answer to this. The things in the iterator need to be closed and can potentially be very large.


https://github.com/NVIDIA/spark-rapids/pull/4588#discussion_r799836101

**Comment by abellina** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by abellina** [comment]
Ok I have a bug with `safeMap`, I thought I had built this before I pushed it last, but obviously I didn't. I am going to retest the patch against real queries to be sure.

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
@jlowe @revans2 ok apologies. Now this should be ready.
